### PR TITLE
fix: preserve buffered headers across better-auth hijack and audit hardening

### PR DESCRIPTION
**Pull request by chuanghiduoc**
## Summary

System-wide audit pass (five parallel review passes across api/common, auth+bootstrap, messaging/workers, DDD modules, and contracts/shared/core/graphql). Every finding below was **verified against live behavior and library source** before fixing; documented-intentional customizations (idempotency plugin, outbox triggers, hijack body-propagation, offset-pagination scaffolding, etc.) were deliberately left as-is.

### 🔴 P0 — CORS/headers lost across the Better Auth hijack (`apps/api/src/main.ts`)
`reply.hijack()` never flushes headers buffered via Fastify's `reply.header()` API, and Better Auth's node handler (`better-call`) only *merges* its own headers onto `reply.raw` on the success path. Result: `@fastify/cors`'s `Access-Control-*` (set in its `onRequest` hook) were **dropped from every successful `/api/auth/*` response**, breaking credentialed cross-origin auth — a topology this app explicitly supports (`CORS_ORIGINS`, cross-origin session cookies).

**Fix:** flush `reply.getHeaders()` onto `reply.raw` just before `reply.hijack()`.

Verified live (before → after):
- Before: `GET /api/auth/get-session` with `Origin` → **no** `access-control-allow-origin`.
- After: `access-control-allow-origin` + `access-control-allow-credentials` present on both `get-session` and `sign-up`. Added an e2e regression test (`auth.e2e-spec.ts`, mirrored CORS + flush into `test-app.ts`).

### 🟠 Defensive ordering — problem-details hook before `@fastify/compress` (`main.ts`)
`applyFastifyProblemDetailsHook` ran *after* compress. Today harmless (compress skips `application/problem+json`), but a latent landmine: any compressible-content-type error body ≥1KB would be gzipped, then the hook would fail to `JSON.parse` the buffer and emit a plaintext body under a stale `Content-Encoding: gzip`. Moved before compress, matching the documented idempotency-hook convention.

### 🟡 DRY — shared `redisFixedWindowIncr` helper (`apps/api/src/common/rate-limit/`)
The credential-stuffing limiter (`main.ts`) and Bull Board's failed-auth budget shipped **byte-identical** fixed-window Redis Lua + parsing. Extracted one unit-tested helper; both call sites now share it.

### 🟡 Latent bug — `buildPageMeta` divisor clamp (`libs/shared`)
`Math.max(1, Math.ceil(total / pageSize))` clamped the *result*, not the divisor: `pageSize: 0` → `Infinity` → `hasNextPage` stuck `true`. Now clamps the divisor, mirroring the already-hardened `toListResponse`.

### 🟢 Observability — upload-verification success-path no-op log (`apps/worker`)
Success path logged a READY flip even when the CAS `updateMany` matched 0 rows (already processed). Now logs a skipped-no-op warning, mirroring the reject path.

### 🟢 Security floor — `valibot >= 1.4.2 <2` (`pnpm-workspace.yaml`)
GHSA-5qjj-4xww-7phc (moderate), pulled transitively via prisma/better-auth. Capped below the next major per the repo's override convention.

### 🟢 Docs/comments
- `compose.prod.yml` healthcheck comment → `/api/v1/health/*` (was missing the prefix; would mislead k8s probe config).
- Moved a stale CORS/CSRF WHY comment to the `enableCors` block.
- Noted the absent WS role-authorization primitive at `RolesGuard`'s ws-bypass (footgun for future `@Roles()` on `@SubscribeMessage`).

## Verification
- `nx affected -t typecheck lint test build` (against `origin/main`) — green.
- `nx run api:e2e` (Testcontainers) — **61 tests pass** (incl. new CORS regression).
- Full `build-dev.sh` docker rebuild + boot — green; smoke test passes.
- Live functional sweep on the rebuilt stack: health (4 probes), Swagger (`/docs`, `/docs-json`), GraphQL (introspection + auth-gated `me`), full auth flow (sign-up → session → users/me), CORS on hijacked responses, sensitive-field redaction, metrics/Bull Board IP guards, RFC 9457 error shape — all pass.

## Notes
No behavior change to the two-tier rate-limit keying, outbox transactionality, or module boundaries. The audit confirmed those (and cursor pagination, CQRS wiring, ws-auth lease/interval cleanup, env-validation parity) correct and consistent with the documented design.

## Summary by CodeRabbit

- **Bug Fixes**
  - Preserved CORS and credential headers for authentication session responses.
  - Improved authentication throttling with consistent fixed-window rate limits.
  - Prevented uploads from being marked ready when verification state has changed.
  - Corrected pagination metadata when page size is zero or invalid.
  - Updated documented production health-check paths to use the versioned API routes.
- **Tests**
  - Added regression coverage for authentication CORS handling and rate-limit edge cases.

### CODE REVIEW

File: apps/api/e2e/test-app.ts [hunk 1]
```diff
@@ -209,6 +218,11 @@ export async function createTestApp(): Promise<TestAppContext> {
     if (req.body !== undefined && (req.raw as unknown as { body?: unknown }).body === undefined) {
       (req.raw as unknown as { body: unknown }).body = req.body;
     }
+    // Mirror main.ts: flush headers buffered by @fastify/cors's onRequest hook onto the raw response
+    // before hijack, otherwise Better Auth's node handler drops them from the reply.
+    for (const [name, value] of Object.entries(reply.getHeaders())) {
+      if (value !== undefined) reply.raw.setHeader(name, value);
+    }
```
[the remark this answers is not in the document]
Comment (chuanghiduoc) [reply apps/api/e2e/test-app.ts]: Addressed in 58f94fd: extracted `flushBufferedReplyHeaders()` into apps/api/src/common/http/ and call it from BOTH main.ts's production `authRouteHandler` and the e2e test app, plus a unit test. The CORS regression test now exercises the exact production flush path, so a regression in main.ts's hijack handling breaks the test.
